### DISCUSSION TITLE
[API Request] Support JSON post for new endpoints ?

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#110] [API requests]
### Type of change

API tweak

### Proposal description

This is just as starting discussion to evaluate if servers / clients would embrace supporting posting JSON for some new endpoints.

This would simplify the definition of complex new APIs like transcoding profiles and filtering, multi sorting and everything we could do with the getAllXXX endpoints for example.

### Backward compatibility

- [ ] No backward compatibility impact.

### API details

Allowing some endpoints to support posting a JSON instead of a very long and complex parameter list that would require specific encoding if we need to pass complex objects in them.

**Comment by epoupon** [comment]
Yes
(Why not?)

**Reply by Tolriq** [reply]
I don't know why not, but better to ask than build proposals rejected because some object.

**Comment by paulijar** [comment]
I have no objections. One pain point in the original Subsonic API is how the list type parameters are handled, e.g. `songId` on the API method `createPlaylist`. The repeated URL argument with the same name is really awkward to handle at least in PHP. Using JSON type arguments in the new API methods would certainly be one way to avoid having this problem in the future.

Stating the obvious, we then need to make sure that the supported JSON contents are very clearly documented, e.g. which properties as mandatory and which are optional and what are the data types of each property.

**Reply by Tolriq** [reply]
We should use JSON Schema for that, it should cover a large part of the definition and would help clients.

**Comment by Tolriq** [comment]
@opensubsonic/servers @opensubsonic/clients This is the first step to move on new endpoints please take time to write a simple yes or no with some details.

**Comment by jeffvli** [comment]
Yes.

As a consumer of the API I welcome any change that makes it simpler to use.

**Comment by deluan** [comment]
Yes, for new endpoints. But IMO we should require a proper JSONSchema or OpenAPI definition doc when introducing the new endpoints, to avoid us getting to the same state we are with the old ones.

**Reply by Tolriq** [reply]
I think you said it was not possible to cover current API with a full OpenAPI ?

I'm only familiar with JSONschema hence why I proposed it above but anything simple to either generate from something or already be easily "diffable" to review PR works.

And it must be something that does not prevent or interfere with the previous endpoints if a client decide to use those definitions to generate client code automatically.

**Reply by deluan** [reply]
Yeah, I though about using OpenAPI only for new endpoints, to describe the payload, it would also allow code generation for the models and validation, but it could be confusing to have only new endpoints in OpenAPI.

On the flip-side, AFAIK JSONSchema only allows one object definition per file. Anyway, I agree, JSONSchema seems to be a better approach here.

What is important is, IMO: We should not add new endpoints without a schema.

**Reply by Tolriq** [reply]
> What is important is, IMO: We should not add new endpoints without a schema.

Yes that's a given.

If you can think about a way to make OpenAPI as a global solution without being weird for the previous endpoints then it's better to have one file to ease code generation.

Else with JSONSchema I would have made 1 schema per endpoint and a global schema with $ref to the rest (but that would more be a directory of schema than something really documenting the API)

**Reply by Tolriq** [reply]
@deluan Asked chatGPT and it propose to use the following to still have a comprehensive OpenAPI but for the endpoints with bad polymorphism that can't be properly documented.

WDTY? Do you still have the OpenAPI test you made previoulsy or do you remember who had started to work on one?


```yaml
get:
  tags:
    - Excluded
  summary: "[Undocumented endpoint]"
  description: |
    This endpoint is intentionally undocumented due to implementation specifics or legacy support. It is listed here for completeness but not fully described.
  responses:
    '200':
      description: Successful but undocumented response
      content:
        application/json:
          schema:
            type: object
```

**Reply by Gr3q** [reply]
Which endpoints have problems?

I'm currently manually building the OpenAPI json [here](https://github.com/Gr3q/opensubsonic-cli/blob/reorganise/docs/opensubsonic-openapi.json) (without xml responses first, OpenSubsonic latest version only).

There are some thing that cannot be properly represented (auth paramater groups, default Empty==ALL) but it's nothing major (so far)

**Reply by Tolriq** [reply]
Some endpoints can return different things depending on the id being album or artists or podcast or track and it's then not possible to properly describe what are the mandatory field and default values for the different cases leading to wrong definition. Same for the auth, there's 3 different auth that each requires different kind of parameters. 

The main selling point of having a full openAPI definition was to be able to clients to autogenerate the API wrappers for their language. 

If there's still a way to properly document things and generate a swagger wrapper for example, then it's probably good enough to not enforce the mandatory fields for now and have most optional.

**Reply by Gr3q** [reply]
> If there's still a way to properly document things and generate a swagger wrapper for example, then it's probably good enough to not enforce the mandatory fields for now and have most optional.

That is the only way to go, see below (and recommended by OpenAPI itself).

> Some endpoints can return different things depending on the id being album or artists or podcast or track and it's then not possible to properly describe what are the mandatory field and default values for the different cases leading to wrong definition.

The response depending on a query parameter cannot be described in OpenAPI (I know ID in the docs is already ambiguous) so the schema can just describe the union of all possible return values if necessary. 

I haven't found the proper distinction for a response in the docs that you are talking about yet but I'm only at endpoints starting with `d`. What I found is that `Child` is definitely very permissive but I can't add more restrictions to the openapi.json than what is in the current docs.

> Same for the auth, there's 3 different auth that each requires different kind of parameters.

[Not possible to describe in OpenAPI 3](https://swagger.io/docs/specification/v3_0/describing-parameters/#parameter-dependencies), will be possible in OpenAPI 4 though. I don't see this a problem with opensubsonic.


Edit: I'm most likely going to open a PR with the OpenAPI spec in this repo when I'm done (hopefully it can be accepted without `xml` - although adding the xml responses should not be that big of a job as long as it's documented)

**Reply by Tolriq** [reply]
The auth param groups would help as you can't mix them and many servers will errors. But again this is mostly an issue for autogenerating the clients code. 
I suppose we can warn that it won't work and it should be used as documentation only.

Edit: For the PR I think it's best if the API is build with different files and not one single massive 10k lines when finished with everything and the examples, ....

**Reply by Gr3q** [reply]
In this repo it can live in different files and the root file almost exclusively contains refs which can be resolved when the full specs need to be built (it can has some caveats with resolving, depending on the library used but that's just implementation) 

I'm building it in one file because as I go:
* I'm testing doc generation
* I'm testing client generation
* Almost everything expects it to be a single file

I can break it up afterwards, but in the end something need to be server that can generate docs/clients conveniently by just pointing to it.

**Reply by Tolriq** [reply]
Yeah, the problem is that I don't see how we can auto generate the large json from the current specs, this means that rebase, merge when multiple things changes in multiple PR are quickly a pain :(

IMO if it's possible having it split in json or yaml and have a task in the build system that generate the global json to serve with swagger would be best.
